pkg/cloudprovider/providers/azure/azure_loadbalancer.go

```diff
@@ -1069,33 +1080,40 @@ func (az *Cloud) reconcileSecurityGroup(clusterName string, service *v1.Service,
-				sharedIndex, sharedRule, sharedRuleFound := findSecurityRuleByName(updatedRules, sharedRuleName)
-				if !sharedRuleFound {
-					klog.V(4).Infof("Expected to find shared rule %s for service %s being deleted, but did not", sharedRuleName, service.Name)
-					return nil, fmt.Errorf("Expected to find shared rule %s for service %s being deleted, but did not", sharedRuleName, service.Name)
-				}
-				if sharedRule.DestinationAddressPrefixes == nil {
-					klog.V(4).Infof("Expected to have array of destinations in shared rule for service %s being deleted, but did not", service.Name)
-					return nil, fmt.Errorf("Expected to have array of destinations in shared rule for service %s being deleted, but did not", service.Name)
-				}
-				existingPrefixes := *sharedRule.DestinationAddressPrefixes
-				addressIndex, found := findIndex(existingPrefixes, destinationIPAddress)
-				if !found {
-					klog.V(4).Infof("Expected to find destination address %s in shared rule %s for service %s being deleted, but did not", destinationIPAddress, sharedRuleName, service.Name)
-					return nil, fmt.Errorf("Expected to find destination address %s in shared rule %s for service %s being deleted, but did not", destinationIPAddress, sharedRuleName, service.Name)
-				}
-				if len(existingPrefixes) == 1 {
-					updatedRules = append(updatedRules[:sharedIndex], updatedRules[sharedIndex+1:]...)
-				} else {
-					newDestinations := append(existingPrefixes[:addressIndex], existingPrefixes[addressIndex+1:]...)
-					sharedRule.DestinationAddressPrefixes = &newDestinations
-					updatedRules[sharedIndex] = sharedRule
-				}
-				dirtySg = true
+			sharedRuleName := ""
+			if len(expectedSourceAddressPrefixes) > 1 {
+				sharedRuleName = az.getSecurityRuleName(service, port, strings.Join(expectedSourceAddressPrefixes, "-"))
+			} else {
+				sharedRuleName = az.getSecurityRuleName(service, port, expectedSourceAddressPrefix)
+			}
+			sharedIndex, sharedRule, sharedRuleFound := findSecurityRuleByName(updatedRules, sharedRuleName)
+			if !sharedRuleFound {
+				klog.V(2).Infof("Expected to find shared rule %s for service %s being deleted, but did not", sharedRuleName, service.Name)
+				return nil, fmt.Errorf("Expected to find shared rule %s for service %s being deleted, but did not", sharedRuleName, service.Name)
+			}
+			if sharedRule.DestinationAddressPrefixes == nil {
+				klog.V(2).Infof("Expected to have array of destinations in shared rule for service %s being deleted, but did not", service.Name)
+				return nil, fmt.Errorf("Expected to have array of destinations in shared rule for service %s being deleted, but did not", service.Name)
 			}
+			existingPrefixes := *sharedRule.DestinationAddressPrefixes
+			addressIndex, found := findIndex(existingPrefixes, destinationIPAddress)
+			if !found {
+				klog.V(2).Infof("Expected to find destination address %s in shared rule %s for service %s being deleted, but did not", destinationIPAddress, sharedRuleName, service.Name)
```

Comment:
Errorf

Reply:
https://github.com/kubernetes/kubernetes/blob/f13ae7cf1121133f75db6bffff9f311dabba1f75/pkg/cloudprovider/providers/azure/azure_loadbalancer.go#L1096-L1097

Errorf is right below this line

Reply:
if Error is returned and ultimately shows in the log or events, then we don't need to print earlier, otherwise it will print twice.